src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1) / val);
+        write_volatile(ptr2, read_volatile(ptr2) / val);
+    }
+}
+
+fn write_or(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1) | val);
+        write_volatile(ptr2, read_volatile(ptr2) | val);
+    }
+}
+
+fn write_and(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1) & val);
+        write_volatile(ptr2, read_volatile(ptr2) & val);
+    }
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the result of adding a random value to the index of iteration
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_seq_inc(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestSeqInc");
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let expected_iter =
+        u64::try_from(half_count * 2).context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+
+    let value: usize = random();
+    for i in 0..half_count {
+        timeout_checker.check()?;
+        write_volatile(base_ptr.add(i), value.wrapping_add(i));
+        write_volatile(half_ptr.add(i), value.wrapping_add(i));
+    }
+    compare_regions(base_ptr, half_ptr, half_count, timeout_checker)
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write to all bytes with either 0xff or 0x0 in an alternating pattern
+/// Read and compare the two halves of the memory region
+/// This procedure is repeated 64 times
+pub unsafe fn test_solid_bits(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestSolidBits");
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let expected_iter = u64::try_from(
+        (half_count * 2)
+            .checked_mul(64)
+            .context("expected_iter overflowed")?,
+    )
+    .context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+
+    for i in 0..64 {
+        let val = if i % 2 == 0 { !0 } else { 0 };
+        for j in 0..half_count {
+            timeout_checker.check()?;
+            let val = if j % 2 == 0 { val } else { !val };
+            write_volatile(base_ptr.add(j), val);
+            write_volatile(half_ptr.add(j), val);
+        }
+        compare_regions(base_ptr, half_ptr, half_count, timeout_checker)?;
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write to all bytes with either 0x55 or 0xbb in an alternating pattern
```

Comment:
I think you mean 0x55 and 0xaa? And maybe put a quick explanation for why those two numbers are not just magic numbers :)